### ISSUE TITLE
HTTP2: Client and server disagree on active stream count

### ISSUE THREAD

**Issue by Tratcher** [dotnet/corefx#40663]
_From @JamesNK on Saturday, August 24, 2019 2:24:08 AM_

Issue found during HTTP2 stress testing. One HttpClient is sending multiple requests in parallel, then canceling them via RST_STREAM sent to the server. For an unknown reason the server and client disagree on the active stream count, and the sever eventually sends RST_STREAM (REFUSED_STREAM) and GOAWAY (STREAM_CLOSED). This issue eventually leads to https://github.com/aspnet/AspNetCore/issues/13405.

Client and server should agree on the active stream count and the client should wait to send new requests, without error.

Wireshark log: https://www.dropbox.com/s/5ndspth4t0qjyj6/clientreset-error.zip?dl=0

Exception on the client:

```
Unhandled exception. System.Net.Http.HttpRequestException: An error occurred while sending the request.
---> System.IO.IOException: The request was aborted.
---> System.IO.IOException: The response ended prematurely, with at least 9 additional bytes expected.
   at System.Net.Http.Http2Connection.ReadAtLeastAsync(Stream stream, Memory`1 buffer, Int32 minReadBytes)
   at System.Net.Http.Http2Connection.EnsureIncomingBytesAsync(Int32 minReadBytes)
   at System.Net.Http.Http2Connection.ReadFrameAsync(Boolean initialFrame)
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
   at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
   at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Grpc.Net.Client.Internal.GrpcCall`2.SendAsync(HttpClient client, HttpRequestMessage message)
   at Grpc.Net.Client.Internal.GrpcCall`2.GetResponseHeadersAsync()
```

Repo is currently not public. Contact @JamesNK or @Tratcher for the repo source code and instruction.

_Copied from original issue: aspnet/AspNetCore#13406_

**Comment by Tratcher** [comment]
_From @jkotalik on Saturday, August 24, 2019 7:45:51 PM_

So https://github.com/aspnet/AspNetCore/pull/12704 was supposed to fix this issue. There may be a case we aren't covering.

**Comment by Tratcher** [comment]
_From @Tratcher on Saturday, August 24, 2019 8:10:53 PM_

Indeed. #12704 was focused on the server terminating the request, but the new repro is focused on client aborts. A quick look through the code shows we did account for client RSTs, but maybe we missed something else.

**Comment by Tratcher** [comment]
_From @JamesNK on Monday, August 26, 2019 11:07:11 PM_

This is a bug but not a critical one. 3.1 fix?

**Comment by Tratcher** [comment]
_From @Tratcher on Wednesday, August 28, 2019 10:49:15 PM_

Theory: The client has a race where it can cancel a stream, decrement its count, but not send a RST_STREAM. The server still considers the stream open and you exceed the stream limit.

Repro logs. I set the connection limit to 1, send two requests, and cancel one of them (the delay=100 request).
```
// Good:
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" received HEADERS frame for stream ID 2545 with length 31 and flags END_STREAM, END_HEADERS
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/2 GET http://localhost:5005/?delay=100
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" received RST_STREAM frame for stream ID 2545 with length 4 and flags 0x0
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" received HEADERS frame for stream ID 2547 with length 29 and flags END_STREAM, END_HEADERS
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 111.1166ms 200
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/2 GET http://localhost:5005/?delay=0
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" sending HEADERS frame for stream ID 2547 with length 76 and flags END_HEADERS
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" sending DATA frame for stream ID 2547 with length 21 and flags NONE
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" sending HEADERS frame for stream ID 2547 with length 27 and flags END_STREAM, END_HEADERS
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 0.2761ms 200
// Bad:
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" received HEADERS frame for stream ID 2549 with length 31 and flags END_STREAM, END_HEADERS
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/2 GET http://localhost:5005/?delay=100
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" received HEADERS frame for stream ID 2551 with length 29 and flags END_STREAM, END_HEADERS
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" sending HEADERS frame for stream ID 2549 with length 76 and flags END_HEADERS
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" sending DATA frame for stream ID 2549 with length 23 and flags NONE
trce: Microsoft.AspNetCore.Server.Kestrel[37]
      Connection id "0HLPC0DDKPENV" sending HEADERS frame for stream ID 2549 with length 33 and flags END_STREAM, END_HEADERS
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 1106537.7128ms 200
```

```

            AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);
            using var client = new HttpClient
            {
                BaseAddress = new Uri("http://localhost:5005"),
                DefaultRequestVersion = HttpVersion.Version20
            };
            for (int i = 0; i < 1000; i++)
            {
                // Send a second one that will cancel
                var r2Task = client.GetAsync("?delay=100", new CancellationTokenSource(50).Token);

                // Send one request normally
                var r1Task = client.GetAsync("?delay=0");

                try
                {
                    var r1 = await r1Task;
                    r1.EnsureSuccessStatusCode();
                }
                catch (Exception ex)
                {
                    Console.WriteLine("R1 fail: " + ex.Message);
                }

                try
                {
                    var r2 = await r2Task;
                    r2.EnsureSuccessStatusCode();
                }
                catch (OperationCanceledException)
                {
                    Console.WriteLine("R2 canceled");
                }
                catch (Exception ex)
                {
                    Console.WriteLine("R2 fail: " + ex);
                }
            }
```
```
            app.Run(async context =>
            {
                var delay = context.Request.Query["delay"].ToString();
                if (!string.IsNullOrEmpty(delay))
                {
                    await Task.Delay(int.Parse(delay));
                }
                await context.Response.WriteAsync($"Hello World! {context.Request.Protocol} {delay}");
            });
```

**Comment by scalablecory** [comment]
CC @geoffkizer @stephentoub @eiriktsarpalis

**Comment by geoffkizer** [comment]
@Tratcher In your repro, when you say "I set the connection limit to 1" what does this mean? I'm assuming you are configuring Kestrel to set MAX_CONCURRENT_STREAMS to 1, correct? How do you do this?

**Comment by JamesNK** [comment]
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-2.2#maximum-streams-per-connection

**Comment by geoffkizer** [comment]
> Theory: The client has a race where it can cancel a stream, decrement its count, but not send a RST_STREAM.

Yeah, it looks like there is a race here, introduced by the fix for #40180.

When the request is cancelled, and the request body is already complete (as is the case here, since it's a GET), we are now decrementing the stream count before we send the RST_STREAM. This means there's a window where if another request is waiting on stream count, it can start sending before the RST_STREAM is sent. See here: https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L377

We need to defer the call to Complete until after we send the RST_STREAM.

**Comment by geoffkizer** [comment]
There are two other places we call SendReset as well, aside from the one I linked to above. All of these have the same issue and will need to be fixed.

Basically, whenever we call SendReset, we should avoid calling Complete before we do so, and then call it after the SendReset instead.

**Comment by Tratcher** [comment]
> @Tratcher In your repro, when you say "I set the connection limit to 1" what does this mean? I'm assuming you are configuring Kestrel to set MAX_CONCURRENT_STREAMS to 1, correct? How do you do this?

Yes, the streams per connection limit james linked to.

**Comment by eiriktsarpalis** [comment]
FWIW this reproduces immediately once I constrain kestrel MaxStreamsPerConnections in the corefx stress suite.

**Comment by davidsh** [comment]
>Yeah, it looks like there is a race here, introduced by the fix for #40180.

@geoffkizer Will you be able to get a fix into master and then release/3.0?  The fix for #40180 already went into release/3.0. So, if that fix regresses something then we need to fix the fix.

**Comment by geoffkizer** [comment]
Sure, I will put together a fix for it.

@JamesNK Can you send me the original repro?

**Comment by JamesNK** [comment]
[GrpcStress.zip](https://github.com/dotnet/corefx/files/3557411/GrpcStress.zip)

> Run (and pay attention to the listening port)
> Grpcgreeter.exe
>  
> Run (and augment the url if necessary):
> GrpcGreeterClient.exe basic --cancellation-window-ms 1000 --total-requests 10000000 --url http://localhost:5000

**Comment by geoffkizer** [comment]
@jamesnk Thanks. How long does it usually take to repro?

**Comment by JamesNK** [comment]
I just ran it and it repoed within 10 seconds.

```
BasicStress
url: https://localhost:5001
totalRequests: 10000000
concurrent: 1000
enqueueTimeoutMs: 10000
dequeueTimeoutMs: 200
cancellationWindowMs: 1000
connectionLoad: 2147483647
cancellationLowerBoundMs: 0
serverDelayWindowMs: 100
Will connect to https://localhost:5001
Making 1 HttpClients.
Finished BasicStress in 0:00:00.4566555
Unhandled exception. System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: The request was aborted.
 ---> System.IO.IOException: The response ended prematurely, with at least 9 additional bytes expected.
   at System.Net.Http.Http2Connection.ReadAtLeastAsync(Stream stream, Memory`1 buffer, Int32 minReadBytes)
   at System.Net.Http.Http2Connection.EnsureIncomingBytesAsync(Int32 minReadBytes)
   at System.Net.Http.Http2Connection.ReadFrameAsync(Boolean initialFrame)
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
```

**Comment by geoffkizer** [comment]
@jamesnk

When I try to run the repro, I get:

Unhandled exception. System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid according to the validation procedure.

**Comment by JamesNK** [comment]
New documentation Hot Off the Press just for your problem:
https://docs.microsoft.com/en-us/aspnet/core/grpc/troubleshoot?view=aspnetcore-3.0#call-a-grpc-service-with-an-untrustedinvalid-certificate

**Comment by geoffkizer** [comment]
It looks like the client is already doing that, but I'm still getting an SSL error:

```csharp
            HttpClientHandler handler = new HttpClientHandler();
            handler.ServerCertificateCustomValidationCallback = delegate { return true; };
            var httpClient = new HttpClient(handler)
            {
                BaseAddress = new Uri(url),
            };
```

**Comment by JamesNK** [comment]
Hmmm, I don't know then. Just to get this sample working, did you try trusting the ASP.NET Core certificate?

Is there a bug in HttpClient around ignoring invalid certificates in HTTP/2?

**Comment by geoffkizer** [comment]
> Is there a bug in HttpClient around ignoring invalid certificates in HTTP/2?

Not that I'm aware of. Our tests do this.

> Hmmm, I don't know then. Just to get this sample working, did you try trusting the ASP.NET Core certificate?

How do I do that?

**Comment by JamesNK** [comment]
It's in the documentation I linked.

**Comment by JamesNK** [comment]
> It looks like the client is already doing that, but I'm still getting an SSL error:
> 
> ```cs
>             HttpClientHandler handler = new HttpClientHandler();
>             handler.ServerCertificateCustomValidationCallback = delegate { return true; };
>             var httpClient = new HttpClient(handler)
>             {
>                 BaseAddress = new Uri(url),
>             };
> ```

Hmm, where do you see that code? The only place I see HttpClient created is in GetClientFactory, Program.cs, line 259:

```
var buckets = (from i in Enumerable.Range(0, bucketCount) select new HttpClient { BaseAddress = new Uri(url) }).ToList();
```

**Comment by geoffkizer** [comment]
Yeah, you are right. I had an old editor window open and got confused and looked at the wrong code for GrpcGreeterClient.

Apologies for the confusion...

**Comment by geoffkizer** [comment]
I can now get it to fail in about 5-10 seconds, but what I'm seeing is different that what you describe above:

Unhandled exception. System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: The request was aborted.
 ---> System.Net.Http.Http2StreamException: The HTTP/2 server reset the stream. HTTP/2 error code 'ENHANCE_YOUR_CALM' (0xb).

**Comment by geoffkizer** [comment]
@tratcher @jkotalik What would cause Kestrel to send ENHANCE_YOUR_CALM?

**Comment by Tratcher** [comment]
That's what you get when too many requests are cancelled but the app hasn't had a chance to finish cleaning them up yet. We have another thread discussing that.

Did you try the repro code I gave above? It avoided that issue. https://github.com/dotnet/corefx/issues/40663#issuecomment-525953428

**Comment by geoffkizer** [comment]
> Did you try the repro code I gave above?

Haven't yet. I will give it a try.

**Comment by geoffkizer** [comment]
@Tratcher I'm running the repro you posted above, but all I'm seeing is the ENHANCE_YOUR_CALM issue.

It's possible I'm doing something wrong in how I'm configuring Kestrel. Here's the code I used:

```csharp
            WebHost.CreateDefaultBuilder(args)
                .UseStartup<Startup>()
                .ConfigureKestrel((content, options) =>
                {
                    options.Limits.Http2.MaxStreamsPerConnection = 1;
                    options.Listen(IPAddress.Loopback, 5000, listenOptions => { listenOptions.Protocols = HttpProtocols.Http2; });

                });
```

**Comment by geoffkizer** [comment]
I can repro what I think is the original issue here with the HttpStress app. I see occasional "the response ended prematurely" errors here, and one REFUSED_STREAM also.

These are swamped by ENHANCE_YOUR_CALM errors, though. In ~5 minutes, I see 4 "the response ended prematurely" errors vs >200 ENHANCE_YOUR_CALM errors.

@Tratcher Can you link to the issue for ENHANCE_YOUR_CALM? Getting that fixed would make it a lot easier to investigate this...

**Comment by Tratcher** [comment]
ENHANCE_YOUR_CALM isn't something we plan to address for 3.0 or 3.1. The client is considered unhealthy or abusive.
https://github.com/aspnet/AspNetCore/issues/13097

**Comment by geoffkizer** [comment]
@Tratcher Why is the client considered unhealthy here? And what would you want the client to do differently?

My understanding of this issue (please correct me if I'm wrong) is that we're hitting a limit because the *server* has reset too many connections. Is that incorrect?

**Comment by geoffkizer** [comment]
@Tratcher That said, I don't understand why I'm seeing ENHANCE_YOUR_CALM with the repro you provided. There's no server-initiated reset here. Any idea what's up here?

**Comment by Tratcher** [comment]
> @Tratcher Why is the client considered unhealthy here? And what would you want the client to do differently?
>
> My understanding of this issue (please correct me if I'm wrong) is that we're hitting a limit because the _server_ has reset too many streams. Is that incorrect?

@geoffkizer the current repro is for client initiated resets, not server resets.

A client that initiates and cancels too many requests is considered as unhealthy. In this case it's a stress test, why would a real client cancel this many of its requests this frequently?

> @Tratcher That said, I don't understand why I'm seeing ENHANCE_YOUR_CALM with the repro you provided. There's no server-initiated reset here. Any idea what's up here?

You're getting ENHANCE_YOUR_CALM with my repro? I wrote it to limit/avoid those by limiting it to 1 concurrent stream and only sending requests in pairs.

**Comment by geoffkizer** [comment]
> A client that initiates and cancels too many requests is considered as unhealthy by the server. 

Ok, that seems reasonable. So I can understand better, what's the limit that triggers ENHANCE_YOUR_CALM?

It seems like it would be nice to be able to disable this for stress testing. In this case it seems to be swamping other issues. But it's not critical.

> You're getting ENHANCE_YOUR_CALM with my repro? I wrote it to limit/avoid those by limiting it to 1 concurrent stream and only sending requests in pairs.

Yes, I consistently see ENHANCE_YOUR_CALM after ~50 iterations. As a result, I can't seem to repro the original issue here with your repro, because the ENHANCE_YOUR_CALM error always seems to get hit first.

Are you not seeing ENHANCE_YOUR_CALM with your repro? If so, it's possible I'm doing something stupid here. I posted the code I'm using to configure Kestrel above; does that look wrong?

**Comment by Tratcher** [comment]
Your config looks fine. Did you include my Startup code linked above? https://github.com/dotnet/corefx/issues/40663#issuecomment-525953428

It's still racy, you may need to play with the delay values.

The limit for ENHANCE_YOUR_CALM is having more than 2x MaxStreamsPerConnection not fully cleaned up by the server. In this case that would be trying to execute 3 or more requests concurrently on the server, such as cancelling 2 but not waiting for them to finish on the server before starting the next iteration.

**Comment by geoffkizer** [comment]
> Your config looks fine. Did you include my Startup code linked above?

Yes. I copied your code, added the Kestrel config code I posted above, and ran it.

I'll try playing around with it to see if I can get it to behave differently.

> such as cancelling 2 but not waiting for them to finish on the server before starting the next iteration.

I'm not sure what it means to wait for requests to finish on the server, in this case. From the client point of view, we've sent RST_STREAM and we consider the request complete. We can't know if/when the server considers them complete.

**Comment by geoffkizer** [comment]
> Have you fixed the missing client RST yet? ENHANCE_YOUR_CALM might be easier to hit after fixing that.

What missing client RST?

The known issue we have here is that there's an occasional race where we will start the next request (i.e. send HEADERS) before sending the RST_STREAM. But the window here is very small and hard to repro. I've been trying to get a reliable repro for this so I can confirm that the fix actually addresses the repro.

**Comment by Tratcher** [comment]
Yes, that one. In my repro it wasn't just a reversal of the RST and the next HEADERS, the RST was never sent at all.

**Comment by geoffkizer** [comment]
> the RST was never sent at all.

Well, that sounds like a different issue then. 

Unfortunately, I haven't been able to repro anything other than ENHANCE_YOUR_CALM with your repro. Suggestions welcome.

**Comment by geoffkizer** [comment]
@Tratcher When you saw the "missing RST", was this on the very first iteration or did it take a few iterations?

**Comment by Tratcher** [comment]
In my example above it happened on stream 2549, but it took less than a minute to reach that condition.

**Comment by karelz** [comment]
This does not seem to be severe enough to meet the 3.0 bar at this point. Moving to 5.0.
Can someone please clarify what is impact of this but? Is 5.0 sufficient milestone to have it addressed?

**Comment by geoffkizer** [comment]
I identified a race in our cancellation logic (see above) that I think is the root cause here, but I haven't been able to confirm that yet. At the very least, I think we should confirm that.

The impact here is that when you are at the stream limit (default 100), and you cancel a stream, there's a race that will occasionally cause us to start the next stream before cancelling the previous stream. This exceeds the stream limit, so the server sees this as a protocol violation and kills the connection, which will result in failures on other streams in flight. The race window is quite small, and it only happens when you are at the stream limit, so it's not that common.

The errors here are swamped by hitting the cancellation rate-limiting logic in ASP.NET that causes requests to fail with ENHANCE_YOUR_CALM when too many cancellations happen. See https://github.com/aspnet/AspNetCore/issues/13668. That has made this kind of painful to investigate. But it also means that in practice, if you're hitting this issue, you're likely hitting the ENHANCE_YOUR_CALM issue far more often anyway, and that's considered by design for 3.0. So it's not clear we actually get much value from fixing this for 3.0.

So basically:
(a) hitting this is pretty rare
(b) if you are hitting this, you're probably hitting the ENHANCE_YOUR_CALM issue anyway and that's not planned to be fixed for 3.0
(c) but if you do hit it, the impact is pretty bad (protocol violation and connection killed)

I'm planning to get the fix ready today or tomorrow. We can make a final decision on whether to take this or not at that point.

**Comment by karelz** [comment]
@geoffkizer are you actively working on it?

**Comment by geoffkizer** [comment]
I have the fix ready. Still trying to validate it using stress and confirm that we're no longer seeing this.

I'll aim to get a PR out relatively soon.
